src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs

```diff
@@ -0,0 +1,140 @@
+        catch (Exception ex)
+        {
+            sw.Stop();
+            context.Logger.Log($"EF analysis ended after {sw.ElapsedMilliseconds} ms with exception {ex}");
+        }
+
+        var telemetry = AnalysisUtilities.GetTelemetry(EFAnalysis, stats, sw);
+        if (telemetry.ExpressionsFound > 0)
+        {
+            context.Telemetry.EFAnalysisResult(telemetry);
+        }
+
+        return telemetry.ExpressionsFound > 0;
+    }
+
+    private static void ReportInvalidExpressions(MongoAnalysisContext context, ExpressionsAnalysis EFExpressionAnalysis)
+    {
+        var semanticContext = context.SemanticModelAnalysisContext;
+
+        if (EFExpressionAnalysis.InvalidExpressionNodes.EmptyOrNull())
+        {
+            return;
+        }
+
+        var driverVersion = ReferencesProvider.GetMongoDBDriverVersion(semanticContext.SemanticModel.Compilation.References);
+        var driverVersionString = driverVersion?.ToString(3);
+
+        foreach (var invalidLinqNode in EFExpressionAnalysis.InvalidExpressionNodes)
+        {
+            var diagnostics = Diagnostic.Create(
+                DiagnosticsRules.DiagnosticRuleNotSupportedLinqExpression,
+                invalidLinqNode.OriginalExpression.GetLocation(),
+                AnalysisUtilities.DecorateMessage(invalidLinqNode.Errors.FirstOrDefault(), driverVersionString, context.Settings));
+
+            semanticContext.ReportDiagnostic(diagnostics);
+        }
+    }
+
+    private static AnalysisStats ReportMqlOrInvalidExpressions(MongoAnalysisContext context, ExpressionsAnalysis EFExpressionAnalysis)
+    {
+        var semanticContext = context.SemanticModelAnalysisContext;
+
+        if (EFExpressionAnalysis.AnalysisNodeContexts.EmptyOrNull())
+        {
+            return AnalysisStats.Empty;
+        }
+
+        var compilationResult = AnalysisCodeGenerator.Compile(context, EFExpressionAnalysis);
+        if (!compilationResult.Success)
+        {
+            return AnalysisStats.Empty;
+        }
+
+        var driverVersion = compilationResult.LinqTestCodeExecutor.DriverVersion;
+        var settings = context.Settings;
+        int mqlCount = 0, internalExceptionsCount = 0, driverExceptionsCount = 0;
+        var typesMapper = new TypesMapper(
+          MqlGeneratorSyntaxElements.Linq.MqlGeneratorNamespace,
+          context.TypesProcessor.GeneratedTypeToOriginalTypeMapping);
+
+        foreach (var analysisContext in EFExpressionAnalysis.AnalysisNodeContexts)
+        {
+            var mqlResult = compilationResult.LinqTestCodeExecutor.GenerateMql(analysisContext.EvaluationMethodName);
+            var locations = analysisContext.Node.Locations;
+
+            if (mqlResult.Mql != null)
+            {
+                var mql = analysisContext.Node.ConstantsRemapper.RemapConstants(mqlResult.Mql);
+                var diagnosticDescriptor = mqlResult.Linq3Only ? DiagnosticsRules.DiagnosticRuleNotSupportedEFExpression : DiagnosticsRules.DiagnosticRuleEF2MQL;
```

Comment:
No linq2 in ef I think. Verify with Damien?

Reply:
Only LINQ3